src/t_cose_encrypt_dec.c

```diff
@@ -0,0 +1,467 @@
+    struct t_cose_key      cek_key;
+
+    /* Initialize decoder */
+    QCBORDecode_Init(&DC,
+                     (UsefulBufC){cose, cose_len},
+                     QCBOR_DECODE_MODE_NORMAL);
+
+   /* Make sure the first item is a tag */
+    result = QCBORDecode_GetNext(&DC, &Item);
+
+    /* Check whether tag is CBOR_TAG_COSE_ENCRYPT or CBOR_TAG_COSE_ENCRYPT0 */
+    if (QCBORDecode_IsTagged(&DC, &Item, CBOR_TAG_COSE_ENCRYPT) == false &&
+        QCBORDecode_IsTagged(&DC, &Item, CBOR_TAG_COSE_ENCRYPT0) == false) {
+        return(T_COSE_ERR_INCORRECTLY_TAGGED);
+    }
+
+    /* protected header */
+    result = QCBORDecode_GetNext(&DC, &protected_hdr);
+
+    if (result != QCBOR_SUCCESS) {
+        return(T_COSE_ERR_CBOR_FORMATTING);
+    }
+
+    if (protected_hdr.uDataType != QCBOR_TYPE_BYTE_STRING) {
+        return(T_COSE_ERR_PARAMETER_CBOR);
+    }
+
+    /* Re-initialize to parse protected header */
+    QCBORDecode_Init(&DC2,
+                     (UsefulBufC)
+                     {
+                      protected_hdr.val.string.ptr,
+                      protected_hdr.val.string.len
+                     },
+                     QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterMap(&DC2, NULL);
+
+    QCBORDecode_GetInt64InMapN(&DC2, COSE_HEADER_PARAM_ALG, &algorithm_id);
+
+    QCBORDecode_ExitMap(&DC2);
+
+    result = QCBORDecode_Finish(&DC2);
+
+    if (result != QCBOR_SUCCESS) {
+        return(T_COSE_ERR_CBOR_FORMATTING);
+    }
+
+    /* unprotected header */
+    QCBORDecode_EnterMap(&DC, NULL);
+
+    QCBORDecode_GetByteStringInMapN(&DC, COSE_HEADER_PARAM_IV, &nonce_cbor);
+
+    if (QCBORDecode_GetError(&DC) !=0 ) {
+         return(T_COSE_ERR_CBOR_MANDATORY_FIELD_MISSING);
+    }
+
+    if (me->key_distribution == T_COSE_KEY_DISTRIBUTION_DIRECT) {
+        QCBORDecode_GetByteStringInMapN(&DC, COSE_HEADER_PARAM_KID, &kid_cbor);
+
+        if (QCBORDecode_GetError(&DC) !=0 ) {
+             return(T_COSE_ERR_CBOR_MANDATORY_FIELD_MISSING);
+        }
+    }
+
+    QCBORDecode_ExitMap(&DC);
+
+    /* Ciphertext */
+    result = QCBORDecode_GetNext(&DC, &Cipher);
+
+    if (result != QCBOR_SUCCESS) {
+        return(T_COSE_ERR_CBOR_FORMATTING);
+    }
+
+    if (Cipher.val.string.len != 0) {
+        ciphertext = (uint8_t *) Cipher.val.string.ptr;
+        ciphertext_len = Cipher.val.string.len;
+        detached_mode = false;
+    } else {
+        ciphertext = detached_ciphertext;
+        ciphertext_len = detached_ciphertext_len;
+        detached_mode = true;
+    }
+
+    /* Two key distribution mechanisms are supported, namely
+     *  - Direct key distribution (where no recipient info is included)
+     *  - HPKE-based key distribution (which requires recipient info)
+     */
+    if (me->key_distribution == T_COSE_KEY_DISTRIBUTION_DIRECT) {
+        if (kid_cbor.len == 0 ||
+            strncmp(me->kid.ptr, kid_cbor.ptr, me->kid.len) != 0
```

Comment:
Use usefulbuf compare function instead of strncmp. Use of strncmp here is bug because the data is binary, not a string.